llvm/lib/Target/NVPTX/NVPTXInstrInfo.td

```diff
@@ -3917,3 +3903,63 @@ def atomic_thread_fence_seq_cst_cta :
 def atomic_thread_fence_acq_rel_cta :
   NVPTXInst<(outs), (ins), "fence.acq_rel.cta;", []>,
   Requires<[hasPTX<60>, hasSM<70>]>;
+
+def fpimm_positive_zero : FPImmLeaf<fAny, [{
+  return Imm.isExactlyValue(+0.0);
+}]>;
+
+def fpimm_positive_zero_v2f16 : PatFrag<(ops), (v2f16 (bitconvert (i32 0)))>;
+def fpimm_positive_zero_v2bf16 : PatFrag<(ops), (v2bf16 (bitconvert (i32 0)))>;
+
+// Patterns will only be used if FMA has a single use, in order to mitigate register pressure
+def NVPTX_fma_oneuse : PatFrag<(ops node:$a, node:$b, node:$c),
+                                  (fma node:$a, node:$b, node:$c), [{
+  return N->hasOneUse();
+}]>;
+
+def FMARELU_F16 :
+  NVPTXInst<(outs Int16Regs:$dst), (ins Int16Regs:$a, Int16Regs:$b, Int16Regs:$c),
+    "fma.rn.relu.f16 \t$dst, $a, $b, $c;", []>,
+    Requires<[useFP16Math, hasPTX<70>, hasSM<80>]>;
+def FMARELU_BF16 :
+  NVPTXInst<(outs Int16Regs:$dst), (ins Int16Regs:$a, Int16Regs:$b, Int16Regs:$c),
+    "fma.rn.relu.bf16 \t$dst, $a, $b, $c;", []>,
+    Requires<[hasBF16Math, hasPTX<70>, hasSM<80>]>;
+def FMARELU_F16X2 :
+  NVPTXInst<(outs Int32Regs:$dst), (ins Int32Regs:$a, Int32Regs:$b, Int32Regs:$c),
+    "fma.rn.relu.f16x2 \t$dst, $a, $b, $c;", []>,
+    Requires<[useFP16Math, hasPTX<70>, hasSM<80>]>;
+def FMARELU_BF16X2 :
+  NVPTXInst<(outs Int32Regs:$dst), (ins Int32Regs:$a, Int32Regs:$b, Int32Regs:$c),
+    "fma.rn.relu.bf16x2 \t$dst, $a, $b, $c;", []>,
+    Requires<[hasBF16Math, hasPTX<70>, hasSM<80>]>;
+// FTZ variants are only supported by fp16, not bf16
+def FMARELU_F16_FTZ :
+  NVPTXInst<(outs Int16Regs:$dst), (ins Int16Regs:$a, Int16Regs:$b, Int16Regs:$c),
+    "fma.rn.ftz.relu.f16 \t$dst, $a, $b, $c;", []>,
+    Requires<[useFP16Math, hasPTX<70>, hasSM<80>]>;
+def FMARELU_F16X2_FTZ :
+  NVPTXInst<(outs Int32Regs:$dst), (ins Int32Regs:$a, Int32Regs:$b, Int32Regs:$c),
+    "fma.rn.ftz.relu.f16x2 \t$dst, $a, $b, $c;", []>,
+    Requires<[useFP16Math, hasPTX<70>, hasSM<80>]>;
+
+// FTZ variants are only supported by fp16, not bf16
+def : Pat<(f16 (fmaxnum (NVPTX_fma_oneuse Int16Regs:$a, Int16Regs:$b, Int16Regs:$c), fpimm_positive_zero)),
```

Comment:
How does FMA handle NaNs? `fmaxnum` requires `If either operand is a NaN, returns the other non-NaN operand.`

If we pass a NaN into an `fma.relu`, and get back a `NaN` back and that would be incorrect for `fmaxnum`.
The documentation does not seem to say anything about NaN handling by FMA (it only mentions it for .sat variants)

Reply:
We are only applying this pattern if `allowUnsafeFPMath` is enabled. So AFAIA we don't need to worry about NaN propagation.

Reply:
The comment for the `allowUnsafeFPMath` says that it only affects precision of the operations, not NaN handling:
https://github.com/llvm/llvm-project/blob/7f60f1312ae007d645fc96618db1238baea088b9/llvm/include/llvm/Target/TargetOptions.h#L171
So, it would allow using a regular FMA instead of `a*b+c`, but `NaN` propagation would not be affected.
This is not the case for `fmaxnum` -> `fma`, where `fmaxnum` has distinctly different requirements for processing NaN inputs.

In any case, we should run that by someone who's familiar with FP handling nuances. 
@arsenm @RKSimon -- what would be a sufficient constraint to reflect that the instruction is not expected to deal with NaNs?

Reply:
Definitely do not rely on allowUnsafeFPMath for no nans. We really should delete allowUnsafeFPMath. 

FMA propagates nan as normal, and may also produce output nans based on the operands.

I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

Reply:
Thanks, I have changed the constraints to `Requires<[noNaNsFPMath]>`.

> I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

I have extended the fpimm pattern to also accommodate negative zero.

Reply:
> I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

I think I misunderstood what you were saying first time round. You are concerned with a comparison of `0.0` and `-0.0`. Since I am pattern matching with `fmaxnum`, which says either -0.0 or 0.0 may be returned in a comparison of both values, the guarantees of `fma.relu` (which doesn't specify what is returned when signed and unsigned zeros are in the final comparison) are the same as the `fmaxnum` comparison.

Reply:
We are fixing the definition of fmaxnum/fminnum to require correctly ordered signed zeroes. The fuzzy behavior can be achieved with nsz, so it doesn't make sense to define it this way.

Best to respect this now to reduce the amount of code that needs changing

Reply:
I've added a PatFrag to make sure the `fmaxnum` has `nsz`.